[WorkflowCard] Implement WorkflowCard component

manojleaton opened the pull request:
<!-- If this pull request fixes an Issue, link it below. If not, you can remove the line below -->

Fixes #BLUI-5097 .

<!-- Include a bulleted list summarizing the main changes you have made in this PR -->

#### Changes proposed in this Pull Request:

- Added WorkflowCard component
- Updated JS docs
- Added test cases
- Updated the demo in example
- Updated the Readme

<!-- Include screenshots if they will help illustrate the changes in this PR -->

#### Screenshots / Screen Recording (if applicable)
<img width="329" alt="Screenshot 2024-01-26 at 9 35 32 PM" src="https://github.com/etn-ccis/blui-react-native-workflows/assets/130662346/5c8e491c-b4c3-4fd6-9bf5-5be43a7fcfa5">
<img width="910" alt="Screenshot 2024-01-26 at 9 51 31 PM" src="https://github.com/etn-ccis/blui-react-native-workflows/assets/130662346/6f067629-e8b8-47cf-8d76-81e0fe22b48e">


<!-- Instruction for PR reviewers, if more complicated than a simple yarn start -->

#### To Test:

- yarn start:example-ios/android

<!-- Useful for draft pull requests -->

#### Any specific feedback you are looking for?

- Need feedback on rendering svg in the card background, as it is done differently in ReactNative. We have added svg to the background in React workflow.
- For loading do we need to use ProgressBar component?

huayunh commented:
what exactly is happening with that header text being chopped off, and why is the mobile view still rendered as a card? Why is the card not using the safe area on the bottom on mobile?

huayunh commented on login-workflow/src/__tests__/components/WorkflowCard/WorkflowCard-test.tsx:
```diff
@@ -0,0 +1,23 @@
+/**
+ * @format
+ */
+
```
I didn't check the rest of the code... does this comment do anything?

manojleaton replied:
Removed the comment.

huayunh reviewed (commented):
what exactly is happening with that header text being chopped off, and why is the mobile view still rendered as a card? Why is the card not using the safe area on the bottom on mobile?

surajeaton commented on login-workflow/src/components/WorkflowCard/WorkflowCard.tsx:
```diff
@@ -0,0 +1,76 @@
+
+type WorkflowCardHeaderProps = CardTitleProps;
+
+type WorkflowCardActionsProps = CardActionsProps;
```
Please export this type.

surajeaton commented on login-workflow/src/components/WorkflowCard/WorkflowCard.tsx:
```diff
@@ -0,0 +1,76 @@
+    /**
+     * A custom background to render behind the card
+     */
+    backgroundImage?: string;
```
We are not using this `backgroundImage` prop anywhere. We should use that.

surajeaton commented on login-workflow/src/components/WorkflowCard/WorkflowCard.tsx:
```diff
@@ -0,0 +1,76 @@
+    /**
+     * If true, a blocking progress spinner will be displayed over the card
+     */
+    loading?: boolean;
```
We should make use of this prop in workflow card.

manojleaton commented:
> what exactly is happening with that header text being chopped off, and why is the mobile view still rendered as a card? Why is the card not using the safe area on the bottom on mobile?

1. Created a issue for the header being chopped off in Tablet in React Native Component Library,  [Issue](https://github.com/etn-ccis/blui-react-native-component-library/issues/531).
2. Removed the default border radius of the RNP Card component. Please check the screenshot for the updated mobile view.
<img width="320" alt="Screenshot 2024-01-29 at 6 01 42 PM" src="https://github.com/etn-ccis/blui-react-native-workflows/assets/130662346/fcc4880f-3f93-4fdf-a983-c5b382328684">

huayunh commented:
Sorry I just noticed this -- what's happening with the text field? The red underline is on the top?

joebochill commented:
This looks broken when the device is rotated:
<img width="892" alt="image" src="https://github.com/etn-ccis/blui-react-native-workflows/assets/29152776/05ba586c-482d-4c58-912c-682d12817404">

Hardcoding fixed dimension for the card should probably come with a maxWidth/maxHeight too just to be on the safe side.

Looks like you are always rendering a SafeAreaView. Per conversation I had with Arshdeep, we should be using a normal view and conditionally adding top padding depending on the presence of a WorkflowCardHeader child. Are you going to add that here, or is @ArshdeepSingh3-eaton going to make the changes in a follow-up?

joebochill commented:
> Sorry I just noticed this -- what's happening with the text field? The red underline is on the top?

Looks like maybe an issue with the WorkflowCardBody? If I remove it, the input looks fine.

manojleaton commented:
> This looks broken when the device is rotated: <img alt="image" width="892" src="https://private-user-images.githubusercontent.com/29152776/300930713-05ba586c-482d-4c58-912c-682d12817404.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY2OTg4NjMsIm5iZiI6MTcwNjY5ODU2MywicGF0aCI6Ii8yOTE1Mjc3Ni8zMDA5MzA3MTMtMDViYTU4NmMtNDgyZC00YzU4LTkxMmMtNjgyZDEyODE3NDA0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTMxVDEwNTYwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE0NDFlNDgzYTdhZDk1NWJjZmVkMWJkMTRiMWJjY2ExZTgyMjdkZTI4MTFjMmE2N2JmYzFiODVlNjVmMTEzNTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.oAYgTFa7MvYcklzRU4a6wIYr8KeddUj6YeU9LHCBlVc">
> 
> Hardcoding fixed dimension for the card should probably come with a maxWidth/maxHeight too just to be on the safe side.
> 
> Looks like you are always rendering a SafeAreaView. Per conversation I had with Arshdeep, we should be using a normal view and conditionally adding top padding depending on the presence of a WorkflowCardHeader child. Are you going to add that here, or is @ArshdeepSingh3-eaton going to make the changes in a follow-up?

1. I have added maxHeight and maxWidth to the Card, now it's not breaking. 
<img width="710" alt="Screenshot 2024-01-31 at 4 34 54 PM" src="https://github.com/etn-ccis/blui-react-native-workflows/assets/130662346/fc9c134f-5225-4540-9895-6f3e4632817a">


2. I have changed to normal View component and added top padding code in this PR itself as @ArshdeepSingh3-eaton suggested.

huayunh commented:
> This looks broken when the device is rotated:

This raises an interesting issue -- do we really want users to use auth workflow in landscape on a phone? Portrait on a phone and portrait/landscape on a tablet makes sense to me. But landscape mobile login is something I'd not support. I'd lock the device orientation on mobile to portrait.

huayunh commented:
And this -- the workflow card is currently avoiding the safe area on the bottom:
![image](https://github.com/etn-ccis/blui-react-native-workflows/assets/8997218/57d041ae-5d8e-4954-80a4-b0f6fbbde256)

I would have expected that white area to touch the device edge, and that iPhone home slider thing to turn black.

ArshdeepSingh3-eaton commented:
can you add style prop as well

ArshdeepSingh3-eaton commented:
@huayunh i don't think that locking the landscape mode on mobile will help as when this library is used by an end user and they have not locked landscape mode
it could cause issue

surajeaton commented on login-workflow/src/hooks/useScreenDimensions.ts:
```diff
@@ -0,0 +1,20 @@
+ */
+import { useWindowDimensions } from 'react-native';
+
+type ScreenWidthProps = {
```
Please rename type to ScreenDimensionsProps

surajeaton commented on login-workflow/src/hooks/useScreenDimensions.ts:
```diff
@@ -0,0 +1,20 @@
+import { useWindowDimensions } from 'react-native';
+
+type ScreenWidthProps = {
+    width: number;
```
Also add js doc

surajeaton commented on login-workflow/src/components/WorkflowCard/WorkflowCard.tsx:
```diff
@@ -0,0 +1,116 @@
+        },
+    });
+
+export type WorkflowCardHeaderProps = CardTitleProps;
```
Please move all the type definitions to types.ts file

ektaghag-eaton reviewed (approved):
Working fine.
